File: src/main/java/com/monovore/hunger/GroupClient.java
//package com.monovore.hunger;
//
//import org.apache.kafka.clients.consumer.OffsetAndMetadata;
//import org.apache.kafka.clients.consumer.internals.ConsumerProtocol;
//import org.apache.kafka.clients.consumer.internals.PartitionAssignor;
//import org.apache.kafka.clients.consumer.internals.PartitionAssignor.Assignment;
//import org.apache.kafka.common.TopicPartition;
//import org.apache.kafka.common.protocol.Errors;
//import org.apache.kafka.common.requests.*;
//import org.slf4j.Logger;
//import org.slf4j.LoggerFactory;
//
//import java.nio.ByteBuffer;
//import java.util.*;
//import java.util.concurrent.CompletableFuture;
//import java.util.stream.Collectors;
//
//public class GroupClient<A> {
//
//    private static final Logger log = LoggerFactory.getLogger(GroupClient.class);
//
//    public A assignment() {
//        return assignment;
//    }
//
//    static class Meta {
//        public final String groupId;
//        public final String protocolType;
//        public final int sessionTimeout;
//
//        public Meta(String groupId, String protocolType, int sessionTimeout) {
//            this.groupId = groupId;
//            this.protocolType = protocolType;
//            this.sessionTimeout = sessionTimeout;
//        }
//
//        public Meta(String groupId, int sessionTimeout) {
//            this(groupId, ConsumerProtocol.PROTOCOL_TYPE, sessionTimeout);
//        }
//    }
//
//    private final ApiClient api;
//    private final Meta meta;
//    private final int generationId;
//    private final String memberId;
//    private A assignment;
//
//    public GroupClient(ApiClient api, Meta meta, A assignment) {
//        this(api, meta, OffsetCommitRequest.DEFAULT_GENERATION_ID, OffsetCommitRequest.DEFAULT_MEMBER_ID, assignment);
//    }
//
//    public GroupClient(ApiClient api, Meta meta, int generationId, String memberId, A assignment) {
//        this.api = api;
//        this.meta = meta;
//        this.generationId = generationId;
//        this.memberId = memberId;
//        this.assignment = assignment;
//    }
//
//    public CompletableFuture<Void> offsetCommit(
//            Map<TopicPartition, OffsetAndMetadata> offsets
//    ) {
//        OffsetCommitRequest request = new OffsetCommitRequest(
//                meta.groupId,
//                generationId,
//                memberId,
//                OffsetCommitRequest.DEFAULT_RETENTION_TIME,
//                offsets.entrySet().stream()
//                        .collect(Collectors.toMap(
//                                Map.Entry::getKey,
//                                entry -> new OffsetCommitRequest.PartitionData(entry.getValue().offset(), entry.getValue().metadata())
//                        ))
//        );
//
//        return api.offsetCommit(request)
//                .thenCompose(results -> {
//
//                    for (short code : results.responseData().values()) {
//                        Errors error = Errors.forCode(code);
//                        if (error != Errors.NONE) return Futures.exceptionalFuture(error.exception());
//                    }
//
//                    return CompletableFuture.completedFuture(null);
//                });
//    }
//
//    public CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> offsetFetch(
//            Set<TopicPartition> partitions
//    ) {
//        OffsetFetchRequest request = new OffsetFetchRequest(meta.groupId, new ArrayList<TopicPartition>(partitions));
//
//        return api.offsetFetch(request).thenCompose(results -> {
//
//            for (OffsetFetchResponse.PartitionData data : results.responseData().values()) {
//                Errors error = Errors.forCode(data.errorCode);
//                if (error != Errors.NONE) return Futures.exceptionalFuture(error.exception());
//            }
//
//            Map<TopicPartition, OffsetAndMetadata> result = results.responseData().entrySet().stream()
//                    .collect(Collectors.toMap(
//                            Map.Entry::getKey,
//                            entry -> new OffsetAndMetadata(entry.getValue().offset, entry.getValue().metadata)
//                    ));
//
//            return CompletableFuture.completedFuture(result);
//        });
//    }
//
//    public CompletableFuture<Void> heartbeat() {
//        HeartbeatRequest request = new HeartbeatRequest(this.meta.groupId, this.generationId, this.memberId);
//        return Futures.lifting(HeartbeatResponse::errorCode, api.heartbeat(request))
//                .thenApply(response -> null);
//    }
//
//    public CompletableFuture<GroupClient<Void>> leave() {
//        log.info("Member '{}' leaving group {}", memberId, meta.groupId);
//        LeaveGroupRequest request = new LeaveGroupRequest(this.meta.groupId, this.memberId);
//        return Futures.lifting(LeaveGroupResponse::errorCode, api.leaveGroup(request))
//                .thenApply(response -> new GroupClient<Void>(api, meta, null));
//    }
//
//    /**
//     * Join or rejoin a group.
//     *
//     * Once this method is called, the current client becomes invalid and should be discarded.
//     *
//     * @param topics The set of topics we want to subscribe to.
//     * @param assignors A list of possible partition-assignment strategies, in order of preference.
//     * @return a new client that has successfully joined the group, or an error.
//     */
//    public CompletableFuture<GroupClient<Assignment>> join(Set<String> topics, List<PartitionAssignor> assignors) {
//
//        List<JoinGroupRequest.ProtocolMetadata> protocols =
//                assignors.stream()
//                        .map(assignor ->
//                                new JoinGroupRequest.ProtocolMetadata(
//                                        assignor.name(),
//                                        ConsumerProtocol.serializeSubscription(assignor.subscription(topics))
//                                )
//                        )
//                        .collect(Collectors.toList());
//
//        JoinGroupRequest joinRequest =
//                new JoinGroupRequest(meta.groupId, meta.sessionTimeout, this.memberId, meta.protocolType, protocols);
//
//        log.info("Joining group '{}' with member id '{}'...", meta.groupId, memberId);
//
//        return Futures.lifting(JoinGroupResponse::errorCode, api.joinGroup(joinRequest))
//                .thenCompose(joinResponse -> {
//
//                    final CompletableFuture<Map<String, ByteBuffer>> groupAssignment;
//                    if (joinResponse.isLeader()) {
//
//                        log.info("Joined group '{}' as leader; generating assignments and syncing...", meta.groupId);
//
//                        String protocol = joinResponse.groupProtocol();
//                        PartitionAssignor partitionAssignor =
//                                assignors.stream()
//                                        .filter(assignor -> assignor.name().equals(protocol))
//                                        .findFirst()
//                                        .get(); // Guaranteed to return one of the specified protocols here
//
//                        Map<String, PartitionAssignor.Subscription> subscriptions =
//                                joinResponse.members().entrySet().stream()
//                                        .collect(Collectors.toMap(Map.Entry::getKey,
//                                                entry -> ConsumerProtocol.deserializeSubscription(entry.getValue())
//                                        ));
//
//                        List<String> allTopics = subscriptions.values().stream()
//                                .flatMap(subscription -> subscription.topics().stream())
//                                .collect(Collectors.toList());
//
//                        groupAssignment =
//                                api.metadata(new MetadataRequest(allTopics))
//                                        .thenCompose(metadataResponse -> {
//
//                                            for (Errors error : metadataResponse.errors().values()) {
//                                                if (error != Errors.NONE)
//                                                    return Futures.exceptionalFuture(error.exception());
//                                            }
//
//                                            Map<String, ByteBuffer> result = partitionAssignor.assign(metadataResponse.cluster(), subscriptions)
//                                                    .entrySet().stream()
//                                                    .collect(Collectors.toMap(Map.Entry::getKey,
//                                                            entry -> ConsumerProtocol.serializeAssignment(entry.getValue())
//                                                    ));
//
//                                            return CompletableFuture.completedFuture(result);
//                                        });
//                    } else {
//                        log.info("Joined group '{}' as follower; syncing...", meta.groupId);
//                        groupAssignment = CompletableFuture.completedFuture(Collections.emptyMap());
//                    }
//
//                    return groupAssignment.thenCompose(assignments -> {
//                        SyncGroupRequest syncRequest =
//                                new SyncGroupRequest(meta.groupId, joinResponse.generationId(), joinResponse.memberId(), assignments);
//
//                        return Futures.lifting(SyncGroupResponse::errorCode, api.syncGroup(syncRequest))
//                                .thenApply(syncResponse -> {
//                                    log.info("Synced with group '{}'; done.", meta.groupId);
//                                    Assignment assignment =
//                                            ConsumerProtocol.deserializeAssignment(syncResponse.memberAssignment());
//                                    return new GroupClient<>(this.api, this.meta, joinResponse.generationId(), joinResponse.memberId(), assignment);
//                                });
//                    });
//                });
//    }
//}

File: src/main/scala/com/monovore/hunger/GroupClient.scala
package com.monovore.hunger

import java.nio.ByteBuffer

import cats.effect.IO
import com.monovore.hunger.GroupClient.Generation
import org.apache.kafka.clients.consumer.internals.{ConsumerProtocol, PartitionAssignor}
import org.apache.kafka.common.{Node, TopicPartition}

import scala.collection.JavaConverters._
import org.apache.kafka.common.requests._

object GroupClient {

  case class Generation(coordinator: Node, memberId: String, generationId: Int)

  trait Assignor {
    def name: String
    def subscription: ByteBuffer
    def membership(members: Map[String, ByteBuffer]): IO[Map[String, ByteBuffer]]
    def action(generation: Generation, assignment: ByteBuffer): IO[Unit]
  }

  class Partitioned(
    client: AsyncClient,
    assignor: PartitionAssignor,
    topics: Set[String],
    assignmentAction: PartitionAssignor.Assignment => IO[Unit]
  ) extends Assignor {

    def name: String = assignor.name

    def subscription: ByteBuffer = ConsumerProtocol.serializeSubscription(assignor.subscription(topics.asJava))

    def membership(members: Map[String, ByteBuffer]): IO[Map[String, ByteBuffer]] = {
      for {
        meta <- client.sendUnchecked(new MetadataRequest.Builder(topics.toList.asJava, false))
      } yield {
        val deserialized = members.mapValues(ConsumerProtocol.deserializeSubscription).toList.toMap
        val assignment = assignor.assign(meta.cluster(), deserialized.asJava).asScala
        assignment.mapValues(ConsumerProtocol.serializeAssignment).toList.toMap
      }
    }

    override def action(generation: Generation, assignment: ByteBuffer): IO[Unit] = {
      assignmentAction(ConsumerProtocol.deserializeAssignment(assignment))
    }
  }
}

case class GroupClient(client: AsyncClient, groupId: String) {

  def heartbeat(generation: Generation): IO[Unit] =
    client.send(generation.coordinator, new HeartbeatRequest.Builder(groupId, generation.generationId, generation.memberId))
      .map { _ => () }

  def commitOffsets(generation: Generation, offsets: Map[TopicPartition, OffsetCommitRequest.PartitionData]): IO[Unit] = {
    val request = new OffsetCommitRequest.Builder(groupId, offsets.asJava)
    request.setGenerationId(generation.generationId)
    request.setMemberId(generation.memberId)
    client.send(generation.coordinator, request).map { _ => () }
  }

  def runGroup(assignor: GroupClient.Assignor): IO[Unit] = {

    for {
      response <- KafkaUtils.retryForever {
        client.send(new FindCoordinatorRequest.Builder(FindCoordinatorRequest.CoordinatorType.GROUP, groupId))
      }
      coordinator = response.node
      response <-
        client.send(coordinator,
          new JoinGroupRequest.Builder(
            groupId, 10000, JoinGroupRequest.UNKNOWN_MEMBER_ID, "consumer", List(new JoinGroupRequest.ProtocolMetadata(assignor.name, assignor.subscription)).asJava
          )
        )
      generationId = response.generationId
      memberId = response.memberId
      membership <-
        if (!response.members.isEmpty) {
          // I'm the leader!
          assignor.membership(response.members.asScala.toMap)
        }
        else IO.pure(Map.empty[String, ByteBuffer])
      response <- client.send(coordinator,
        new SyncGroupRequest.Builder(groupId, generationId, memberId, membership.asJava)
      )
      result <- assignor.action(GroupClient.Generation(coordinator, memberId, generationId), response.memberAssignment)
      leave <- client.sendUnchecked(coordinator,
        new LeaveGroupRequest.Builder(groupId, memberId)
      )
    } yield result
  }
}
